File: python/two-sum-01.py
# https://leetcode.com/problems/two-sum/
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        num_map = {} # Using a dictionary as a hash table for speed
        
        for i in range(len(nums)):
            pair_num = target - nums[i]

            if pair_num in num_map:
                return [num_map.get(pair_num), i]
            
            num_map[nums[i]] = i

# # Local Test
# import timeit

# def local_test():
#         run_obj = Solution()
#         run_obj.twoSum([3,2,3], 6)

# def time_test():
#     #trials = 10_000_000
#     trials = 1_000_000
#     kwargs = {'setup' : 'x=42',
#               'globals' : globals(), 
#               'number': trials}

#     time_output = timeit.timeit('local_test()', **kwargs)
#     print(f'{time_output=:.02f}')

# if __name__ == '__main__':
#     #local_test() # expected [0,2]
#     time_test()
    

# Results
# 08/08/2021 20:59	Accepted	40 ms	14.4 MB	python
# 08/21/2021 09:34	Accepted	72 ms	14.4 MB	python (54.38%)
# Solution: Hashmaps

File: javascript/two-sum-01.js
/**
 * https://leetcode.com/problems/two-sum/
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  let numbers = new Map();

  for (i = 0; i < nums.length - 1; i++) {
    pair_num = target - nums[i];

    if (numbers.has(pair_num)) {
      return [numbers.get(pair_num), i];
    }
    numbers.set(nums[i], i);
  }

  return numbers;
};

// Local Test
//let output = twoSum([2,7,11,15], 9);
//let output = twoSum([3,2,4], 6);
//console.log(output);

// Results
// 73.97% - 08/18/2021 16:23	Accepted	76 ms	41.2 MB	javascript
// Solution: Hashmaps
